Packs/Base/Scripts/DBotFindSimilarIncidentsByIndicators/DBotFindSimilarIncidentsByIndicators.py

```diff
@@ -0,0 +1,507 @@
+    """
+    TFIDF class for indicator
+    """
+
+    def __init__(self, feature_names, tfidf_params, normalize_function, x):
+        self.feature_names = feature_names
+        self.params = tfidf_params
+        self.normalize_function = normalize_function
+        self.frequency = {}
+        if self.normalize_function:
+            x = x[self.feature_names].apply(self.normalize_function)
+        else:
+            x = x[self.feature_names]
+        self.vocabulary = x.iloc[0].split(' ')
+
+    def fit(self, x):
+        if self.normalize_function:
+            x = x[self.feature_names].apply(self.normalize_function)
+        else:
+            x = x[self.feature_names]
+        size = len(x) + 1
+        frequencies = Counter(flatten([t.split(' ') for t in x.values]))
+        frequencies.update(Counter(self.vocabulary))
+        self.frequency = {k: math.log(1 + size / v) for k, v in frequencies.items()}
+        return self
+
+    def transform(self, x):
+        if self.normalize_function:
+            x = x[self.feature_names].apply(self.normalize_function)
+        else:
+            x = x[self.feature_names]
+        return x.apply(self.compute_term_score)
+
+    def compute_term_score(self, x):
+        x = x.split(' ')
+        return sum([1 * self.frequency[word] for word in self.vocabulary if word in x]) / sum(
+            [self.frequency[word] for word in self.vocabulary])
+
+
+TRANSFORMATION = {
+    'indicators': {'transformer': Tfidf,
+                   'normalize': None,
+                   'params': {'analyzer': 'word', 'max_features': 200, 'token_pattern': '.'},  # [\d\D]*
+                   'scoring': {'scoring_function': identity_score, 'min': 0.5}
+                   }
+}
+
+
+class Transformer():
+    def __init__(self, p_type, field, p_incidents_df, p_incident_to_match, p_params):
+        self.type = p_type
+        self.field = field
+        self.incident_to_match = p_incident_to_match
+        self.incidents_df = p_incidents_df
+        self.params = p_params
+
+    def fit_transform(self):
+        transformation = self.params[self.type]
+        transformer = transformation['transformer'](self.field, transformation['params'], transformation['normalize'],
+                                                    self.incident_to_match)
+        X_vect = transformer.fit_transform(self.incidents_df)
+        incident_vect = transformer.transform(self.incident_to_match)
+
+        return X_vect, incident_vect
+
+    def get_score(self):
+        scoring_function = self.params[self.type]['scoring']['scoring_function']
+        X_vect, incident_vect = self.fit_transform()
+        dist = scoring_function(X_vect)
+        self.incidents_df['similarity %s' % self.field] = np.round(dist, 2)
+        return self.incidents_df
+
+
+class Model:
+    def __init__(self, p_transformation):
+        self.transformation = p_transformation
+
+    def init_prediction(self, p_incident_to_match, p_incidents_df, p_indicators=[]):
```

Comment:
why we need this function and not using the class constructor?

Reply:
Same logic than the MTH model
Constructor initiate the model with the parameters that we give him
init_prediction will create the framework for this incident only
So we can use the same object to predict on multiple incidents